### PR TITLE
Add libgdiplus dependency

### PR DESCRIPTION
**Pull request by mudler**
- Make the osc reset ignore patches if required
- Skip checksum verification for libgdiplus

Signed-off-by: Christian Richter <crichter@suse.com>

### CODE REVIEW

File: lib/cf_obs_binary_builder/checksum.rb [hunk 1]
```diff
@@ -121,4 +128,9 @@ def self.libmemcache(version)
   def self.scm_dependency(version)
     # noop - the sources are fetched from an scm repository like hg
   end
+
+  def self.libgdiplus(version)
+    # noop - the sources are fetched from a GitHub repository
+  end
+
```
Comment (thardeck) [inline lib/cf_obs_binary_builder/checksum.rb]: I would remove the empty line

File: lib/cf_obs_binary_builder/dependencies/libgdiplus.rb [hunk 2]
```diff
@@ -0,0 +1,33 @@
+class CfObsBinaryBuilder::Libgdiplus < CfObsBinaryBuilder::BaseDependency
+    def initialize(version)
+      super(
+        "libgdiplus",
+        version,
+        "https://github.com/mono/libgdiplus/archive/#{version}.tar.gz"
```
Comment (thardeck) [inline lib/cf_obs_binary_builder/dependencies/libgdiplus.rb]: The version is not fetched from a git repo, we can verify a checksum. And depwatcher has support for it afaik.

File: lib/cf_obs_binary_builder/checksum.rb [hunk 3]
```diff
@@ -48,6 +48,12 @@ class CfObsBinaryBuilder::Checksum
         "extension" => ".tar.gz"
       }
     },
+    "libgdiplus" => {
+      "type" => "github_tags",
+      "params" => {
+        "repo" => "mono/libgdiplus"
+      }
+    },
```
Comment (thardeck) [inline lib/cf_obs_binary_builder/checksum.rb]: This means depwatcher has support for verification.

File: lib/cf_obs_binary_builder/checksum.rb [hunk 4]
```diff
@@ -69,6 +75,7 @@ def self.for(name, version)
     return libffi(version) if name == "libffi"
     return libmemcache(version) if name == "libmemcache"
     return scm_dependency(version) if name == "openjdk"
+    return libgdiplus(version) if name == "libgdiplus"
```
Comment (thardeck) [inline lib/cf_obs_binary_builder/checksum.rb]: Only needed for special cases like mercury synchronizations.

File: lib/cf_obs_binary_builder/templates/libgdiplus.spec.erb [hunk 5]
```diff
@@ -0,0 +1,79 @@
...
+BuildRequires:  pkg-config
+BuildRequires:  pkgconfig(cairo)
+BuildRequires:  pkgconfig(fontconfig)
+BuildRequires:  pkgconfig(freetype2)
+BuildRequires:  pkgconfig(glib-2.0)
+BuildRequires:  pkgconfig(libexif)
+BuildRequires:  pkgconfig(libpng)
+BuildRequires:  pkgconfig(libtiff-4)
+BuildRequires:  pkgconfig(x11)
+BuildRequires:  pkgconfig(xrender)
```
Comment (thardeck) [inline lib/cf_obs_binary_builder/templates/libgdiplus.spec.erb]: The list seems quite long, non of these packages are part of the stack meta package?
https://build.opensuse.org/package/view_file/Cloud:Platform:Stack-SLE:packages/aaa_stack_build_requires/aaa_stack_build_requires.spec?expand=1

File: lib/cf_obs_binary_builder/templates/libgdiplus.spec.erb [hunk 6]
```diff
@@ -0,0 +1,79 @@
...
+<%= rpm_macros %>
+
+%description
+Mono library that provide a GDI+ comptible API on non-Windows
+operating systems.
+
+%prep
+%setup -q -n %{name}
+%patch -p1
+sed -i 's,-Wall,-Wall %{optflags},' src/Makefile.am
```
Comment (thardeck) [inline lib/cf_obs_binary_builder/templates/libgdiplus.spec.erb]: A comment should explain why we do this replacement.

File: lib/cf_obs_binary_builder/templates/libgdiplus.spec.erb [hunk 7]
```diff
@@ -0,0 +1,79 @@
...
+
+%build
+autoreconf -fiv
+#%configure
+./configure --prefix=/ \
+    --enable-static
+make %{?_smp_mflags}
+make DESTDIR=%{dependencydir} install %{?_smp_mflags}
+rm -rf %{buildroot}%{_libdir}/%{name}.{a,la}
```
Comment (thardeck) [inline lib/cf_obs_binary_builder/templates/libgdiplus.spec.erb]: Upstream does not seem to do that: https://github.com/cloudfoundry/buildpacks-ci/blob/master/tasks/build-binary-new/builder.rb#L60

Did you compare our dependency result with theirs?

File: lib/cf_obs_binary_builder/obs_package.rb [hunk 8]
```diff
@@ -36,7 +36,12 @@ def checkout(config = {}, &block)
         Dir.chdir(name) do
           if config[:reset]
             log 'Resetting package'
-            FileUtils.rm Dir.glob("*")
+            if config[:keep_patches]
```
Comment (thardeck) [inline lib/cf_obs_binary_builder/obs_package.rb]: I would prefer if the patch comes out of cf-obs-binary-builder so we do not need a special handling and can also remove it if it is obsolete.

Comment (thardeck) [reply lib/cf_obs_binary_builder/obs_package.rb]: Or maybe even a sed statement is possible if it is a simple patch.

File: lib/cf_obs_binary_builder/checksum.rb (hunk 3 above)
Comment (mudler) [reply lib/cf_obs_binary_builder/checksum.rb]: Not for libgdiplus:
```
{"source":{"name":"libgdiplus","type":"github_tags","repo":"mono/libgdiplus"},"version":{"ref":"5.6"}}
	
{"ref":"5.6","url":"https://github.com/mono/libgdiplus","git_commit_sha":"f07965ec5fc2b22bfb300d5fef410d25ae472a34"}
```

See [Trello card](https://trello.com/c/PWvmQNzj/583-adapt-cf-obs-binary-builder-to-get-rubygems-checksum-from-upstream)

File: lib/cf_obs_binary_builder/templates/libgdiplus.spec.erb (hunk 7 above)
Comment (mudler) [reply lib/cf_obs_binary_builder/templates/libgdiplus.spec.erb]: It's the same:

```
8497636 drwxr-xr-x 1 ettore users    6 Jan 30 13:56 ..
8497640 -rw-r--r-- 1 ettore users 653K Jan 30 11:47 libgdiplus.a
8497644 -rwxr-xr-x 1 ettore users 1.1K Jan 30 11:47 libgdiplus.la
8497643 lrwxrwxrwx 1 ettore users   19 Jan 30 11:47 libgdiplus.so -> libgdiplus.so.0.0.0
8497639 lrwxrwxrwx 1 ettore users   19 Jan 30 11:47 libgdiplus.so.0 -> libgdiplus.so.0.0.0
8497638 -rwxr-xr-x 1 ettore users 483K Jan 30 11:47 libgdiplus.so.0.0.0
8497641 drwxr-xr-x 1 ettore users   26 Jan 30 11:47 pkgconfig
➜  libgdiplus ls -liah out                                                  
➜  libgdiplus ls -liah libgdi_cf 
total 0
8497890 drwxr-xr-x 1 ettore users   6 Jan 30 14:57 .
8497634 drwxr-xr-x 1 ettore users 192 Jan 30 14:57 ..
8497891 drwxr-xr-x 1 ettore users 162 Jan 17 21:42 lib
➜  libgdiplus ls -liah libgdi_cf/lib 
total 5.0M
8497891 drwxr-xr-x 1 ettore users  162 Jan 17 21:42 .
8497890 drwxr-xr-x 1 ettore users    6 Jan 30 14:57 ..
8497896 -rw-r--r-- 1 ettore users 3.2M Jan 17 21:42 libgdiplus.a
8497895 -rwxr-xr-x 1 ettore users 1.1K Jan 17 21:42 libgdiplus.la
8497894 lrwxrwxrwx 1 ettore users   19 Jan 17 21:42 libgdiplus.so -> libgdiplus.so.0.0.0
8497893 lrwxrwxrwx 1 ettore users   19 Jan 17 21:42 libgdiplus.so.0 -> libgdiplus.so.0.0.0
8497892 -rwxr-xr-x 1 ettore users 1.8M Jan 17 21:42 libgdiplus.so.0.0.0
8497897 drwxr-xr-x 1 ettore users   26 Jan 17 21:42 pkgconfig
```

File: lib/cf_obs_binary_builder/obs_package.rb (hunk 8 above)
Comment (mudler) [reply lib/cf_obs_binary_builder/obs_package.rb]: This should be fixed now, patches are under ``templates/patches/``

File: lib/cf_obs_binary_builder/templates/libgdiplus.spec.erb [hunk 9]
```diff
@@ -0,0 +1,79 @@
...
+%prep
+%setup -q -n %{name}
+<% patches.each_with_index do |patch, index| %>
+%patch<%=index%> -p1
+<% end %>
+
+%build
+autoreconf -fiv
+./configure --prefix=/ \
+    --enable-static
```
Comment (thardeck) [inline lib/cf_obs_binary_builder/templates/libgdiplus.spec.erb]: Upstream does not use enable-static, we should also not then.

File: lib/cf_obs_binary_builder/templates/libgdiplus.spec.erb [hunk 10]
```diff
@@ -0,0 +1,79 @@
...
+operating systems.
+
+%prep
+%setup -q -n %{name}
+<% patches.each_with_index do |patch, index| %>
+%patch<%=index%> -p1
+<% end %>
+
+%build
+autoreconf -fiv
```
Comment (thardeck) [inline lib/cf_obs_binary_builder/templates/libgdiplus.spec.erb]: Upstream uses autogen.sh

File: lib/cf_obs_binary_builder/templates/libgdiplus.spec.erb [hunk 11]
```diff
@@ -0,0 +1,77 @@
...
+
+%prep
+%setup -q -n %{name}
+<% patches.each_with_index do |patch, index| %>
+%patch<%=index%> -p1
+<% end %>
+
+%build
+./autogen.sh --prefix=%{dependencydir}
+make
```
Comment (thardeck) [inline lib/cf_obs_binary_builder/templates/libgdiplus.spec.erb]: Use `make %{?_smp_mflags}` instead, this just adds the `-j[0-9]` compiler flag.

**Pull request merged**